### ISSUE TITLE
`Log::AsyncDispatcher` causes `Channel::ClosedError` if message is logged in fiber that never executes

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#13763]
```cr
require "log"

Log.setup :info

spawn do
  Log.info { "in my own fiber" }
end
```
Running this in non-MT mode results in:
```cr
Unhandled exception in spawn(name: game): Channel is closed (Channel::ClosedError)
  from /usr/local/Cellar/crystal/1.9.2/share/crystal/src/channel.cr:227:7 in 'send'
  from /usr/local/Cellar/crystal/1.9.2/share/crystal/src/log/dispatch.cr:55:7 in 'dispatch'
  from /usr/local/Cellar/crystal/1.9.2/share/crystal/src/log/backend.cr:24:5 in 'dispatch'
  from /usr/local/Cellar/crystal/1.9.2/share/crystal/src/log/log.cr:36:3 in '->'
  from /usr/local/Cellar/crystal/1.9.2/share/crystal/src/fiber.cr:146:11 in 'run'
  from /usr/local/Cellar/crystal/1.9.2/share/crystal/src/fiber.cr:98:34 in '->'
```

Ultimately this seems to be related to an [at_exit](https://github.com/crystal-lang/crystal/blob/master/src/log/main.cr#L49) closing all the `Log` related things. Then at some point after that happens, but before the program fully exits, the fiber from the user code executes which ends up trying to log a message again, but the channel is now closed so it fails.

I feel like this isn't desired/expected behavior.

**Comment by straight-shoota** [comment]
`AsyncDispatcher` runs a write loop in a separate fiber. But in this small program there's no chance for a fiber swap, so the write loop only starts *after* its delivery channel is already closed. It can never collect any dispatched messages.
Putting a `Fiber.yield` almost anywhere in the program removes the error in the example.

This is not a very reliable solution because with multiple fibers pending, it gets more complicated. Yielding might still not cause the writer loop to execute properly. Or it could lead to other fibers being awaken and send to the channel, while the channel will be closed before the writer loop can collect. So we'd be back to square one.

We could introduce a mechanism to stop submissions to the dispatch queue, but still allow the writer loop to collect messages that are already submitted. `Channel#close` closes both send and receive actions, so it doesn't work for that.
This of course only moves the error to the log callsite. This would be an improvement because it allows handling the error, which is impossible if it shows up in the internal writer loop.

Anyway, I think the main cause for this issue in the first place is loose fibers. The program exits before fiber `foo` finishes. I would consider this a sloppy concurrency architecture and can result in synchronization errors like this one.
Moving the error to the call site would help pointing this out.

@stakach I presume your use case which prompted #13978 is more complex. Do you have unbounded fibers running and logging?

**Comment by stakach** [comment]
I've only really seen it occur in Specs, I assume as they run in the `at_exit` handler?
Haven't seen it in production but also it's a pain to work around in Specs.

Maybe we could use build flags to prevent it in specs? Is there a build flag to check if running specs?

I am seeing this line of this shard raise an error: https://github.com/petoem/inotify.cr/blob/master/src/inotify/watcher.cr#L42
When used in a spec like:

```crystal
describe MyClass do
  change_count = 0
  Inotify.watch(PIPELINE_CONFIG) do |event|
    change_count += 1
  end
  
  Spec.before_each { change_count = 0 }
  
  it "should do something that triggers a change" {}
end
```

**Comment by straight-shoota** [comment]
Ah, yeah the `at_exit` handlers of `spec` and `log` could indeed be interfeering that it depends on the order of their definition.
I couldn't find a minimal reproduction with just `require "spec"` and `require "log"`. But this example demonstrates the general effect:

```cr
at_exit do
  Log.info { "foo" } # Error running at_exit handler: Channel is closed
end

require "log"

::Log.setup("*", :trace)
```

`at_exit` handlers are executed in the reverse order of their definition.

So what we would need here is a more explicit order that ensures the log handler is always executed after the spec handler. More generally, I think it should be sufficient for the spec handler to execute first.
A very simple approach to that could be to place the call to `Spec.run` in a `macro finished` hook.
```diff
--- i/src/spec.cr
+++ w/src/spec.cr
@@ -132,4 +132,6 @@ Spec.add_split_filter ENV["SPEC_SPLIT"]?
   Process.on_interrupt { Spec.abort! }
 {% end %}

-Spec.run
+macro finished
+  Spec.run
+end
```

**Comment by straight-shoota** [comment]
The issue with `inotify` however seems to be actually an unbounded fiber hanging around:
https://github.com/petoem/inotify.cr/blob/ddc795fd4d65633d4173b8103e6484c45399ebf4/src/inotify/watcher.cr#L35C17-L35C17 
The error trace points to that line:
```
Unhandled exception in spawn: Channel is closed (Channel::ClosedError)
  from /home/linuxbrew/.linuxbrew/Cellar/crystal/1.10.1/share/crystal/src/channel.cr:228:8 in 'send'
  from /home/linuxbrew/.linuxbrew/Cellar/crystal/1.10.1/share/crystal/src/log/dispatch.cr:55:7 in 'dispatch'
  from /home/linuxbrew/.linuxbrew/Cellar/crystal/1.10.1/share/crystal/src/log/backend.cr:24:5 in 'dispatch'
  from /home/linuxbrew/.linuxbrew/Cellar/crystal/1.10.1/share/crystal/src/log/log.cr:36:3 in 'lurk'
  from src/inotify/watcher.cr:35:7 in '->'
  from /home/linuxbrew/.linuxbrew/Cellar/crystal/1.10.1/share/crystal/src/fiber.cr:146:11 in 'run'
  from /home/linuxbrew/.linuxbrew/Cellar/crystal/1.10.1/share/crystal/src/fiber.cr:98:34 in '->'
  from ???
```

**Comment by stakach** [comment]
Of course i want that fiber to hang around, that error occurs before the specs run and hence the specs fail as the file changes are not detected

**Comment by straight-shoota** [comment]
Right, I missed that this already fails before the specs run. Then it's the issue of `at_exit` ordering (https://github.com/crystal-lang/crystal/issues/13763#issuecomment-1812418763).

**Comment by straight-shoota** [comment]
Oh this is really quite a bit confusing. Yes it fails before the specs run. But `Inotify.watch` runs entirely before `Spec.run` because it's in a `describe` block (which executes immediately in contrast to an `it` block). So the `describe` block is really irrelevant for this.

The sequence is actually this:
* Both `Log` and `Spec` register an `at_exit` handler upon being required.
* The top-level code (inside `describe`) calls `Inotify.watch` which raises (at least for me, probably because the monitored file doesn't exist)
* The exception is never captured and pops up in the main routine which causes it to exit the process.
* This triggers the `at_exit` handlers: `Log` shuts down and `Spec` starts running its specs.

The order of exit handler execution doesn't even matter in this case. `Spec` should never execute any examples when the process is already exiting due to a previous error.

**Comment by straight-shoota** [comment]
@stakach This patch ensures the loose worker fiber is properly finished when the watcher closes.
With that I don't get a closed channel error any more.
```diff
--- i/src/inotify/watcher.cr
+++ w/src/inotify/watcher.cr
@@ -30,6 +30,7 @@ module Inotify
       Log.debug { "inotify IO created" }

       @event_channel = Channel(Event).new
+      @lurk_done = Channel(Nil).new
       @enabled = true
       wait_for_event
       spawn lurk
@@ -83,6 +84,8 @@ module Inotify
         # Ignore the `Closed stream (IO::Error)` when watcher is closed
         raise ex
       end
+    ensure
+      @lurk_done.send nil
     end

     private def wait_for_event
@@ -160,6 +163,7 @@ module Inotify
       @enabled = false
       @io.close
       Log.debug { "file descriptor referring to inotify instance closed" }
+      @lurk_done.receive
     end

     def finalize
```

**Comment by stakach** [comment]
The specs work for me if I comment out the call to `Log.debug { "waiting for event data" }` in that fiber
Whilst your patch improves the termination code of the watcher I'm struggling with the initialization code, specifically in specs.

**Comment by stakach** [comment]
can confirm your patch to `spec.cr`
```
macro finished
  Spec.run
end
```
does work for me

**Comment by straight-shoota** [comment]
I'm actually wondering why `spec` even runs in an `at_exit` handler 🤔 Maybe `macro finished` just wasn't available at the time it was introduced and since then it never occured that it could be changed?

Running the spec suite directly from a `macro finished` hook should be pretty much equivalent. It would be semantically cleaner because this is not really an "exit" behaviour: it's the main function of the program! And there'd be no contention with other actual `at_exit` handlers
In exchange, there'd be contention with `macro finished` hooks, but I don't figure that would be much of an issue.
